[Tests] Refactor interface creation for Linux namespaces

MarekPikula opened the pull request:
#### Summary

Changed so that there is clear relationship between up and down commands, with more granular registering of executed commands, allowing for partial cleanup without errors.

Also, added an explicit XMLRPC server interface (`mgmt`), which will be later used for XMLRPC server namespace encapsulation.

#### Related issues

Continuation of work on concurrent tests (@enkiusz, @arkq).

#### Testing

Tested locally using run_test_suite.py

#### Readability checklist

The checklist below will help the reviewer finish PR review in time and keep the
code readable:

-   [x] PR title is
        [descriptive](https://project-chip.github.io/connectedhomeip-doc/contributing/pull_request_guidelines.html#title-formatting)
-   [x] Apply the
        [_“When in Rome…”_](https://project-chip.github.io/connectedhomeip-doc/style/CODING_STYLE_GUIDE.html)
        rule (coding style)
-   [x] PR size is short
-   [x] Try to avoid "squashing" and "force-update" in commit history
-   [x] CI time didn't increase

See: [Pull Request Guidelines](https://project-chip.github.io/connectedhomeip-doc/contributing/pull_request_guidelines.html)

MarekPikula committed:
Refactor interface creation for Linux namespaces

Changed so that there is clear relationship between up and down
commands, with more granular registering of executed commands, allowing
for partial cleanup without errors.

Also, added an explicit XMLRPC server interface, which will be later
used for XMLRPC server namespace encapsulation.

Signed-off-by: Marek Pikuła <m.pikula@partner.samsung.com>

andy31415 reviewed (approved):
Seems ok. @MarekPikula maybe given the amount of classes, it would be worth it to start a separate module for these up/down dataclasses. The linux.py is starting to be very large/complex

MarekPikula committed:
Merge remote-tracking branch 'upstream/master' into refactor-linux-netns

MarekPikula committed:
tests/linux: Remove confusing SubprocessKind.RPC

Signed-off-by: Marek Pikuła <m.pikula@partner.samsung.com>

MarekPikula committed:
tests/linux: Fix network link up method signature

Signed-off-by: Marek Pikuła <m.pikula@partner.samsung.com>

MarekPikula committed:
Merge remote-tracking branch 'upstream/master' into refactor-linux-netns

MarekPikula committed:
tests/linux: Enable IPv6 for all interfaces

Signed-off-by: Marek Pikuła <m.pikula@partner.samsung.com>

MarekPikula committed:
tests/linux: Fix KeyError usage to ValueError

Signed-off-by: Marek Pikuła <m.pikula@partner.samsung.com>

MarekPikula committed:
tests/linux: Fix netns constructor parameter order and type hints

Signed-off-by: Marek Pikuła <m.pikula@partner.samsung.com>

MarekPikula commented on scripts/tests/chiptest/linux.py:
```diff
@@ -68,139 +69,238 @@ def ensure_private_state():
+class IsolatedNetworkNamespace:
+    """Helper class to create and remove network namespaces for tests."""
+
+    def __init__(self, index: int = 0, rpc_link_name='eth-rpc', rpc_link_up: bool = True, app_link_name: str = 'eth-app',
```
Fixed

MarekPikula commented on scripts/tests/chiptest/linux.py:
```diff
@@ -68,139 +69,238 @@ def ensure_private_state():
+            case SubprocessKind.TOOL:
+                return self.tool_ns
+            case _:
+                raise KeyError("Unknown subprocess kind.")
```
Fixed

MarekPikula committed:
Merge remote-tracking branch 'upstream/master' into refactor-linux-netns

Signed-off-by: Marek Pikuła <m.pikula@partner.samsung.com>

MarekPikula committed:
Don't use properties for setup/link_up_cmds

Signed-off-by: Marek Pikuła <m.pikula@partner.samsung.com>

MarekPikula committed:
Don't pass index as link argument

Signed-off-by: Marek Pikuła <m.pikula@partner.samsung.com>

MarekPikula committed:
Rename tool to ctrl and rpc to mgmt

Signed-off-by: Marek Pikuła <m.pikula@partner.samsung.com>

MarekPikula committed:
Refactor the new network resource abstraction

Signed-off-by: Marek Pikuła <m.pikula@partner.samsung.com>

MarekPikula committed:
Merge remote-tracking branch 'upstream/master' into refactor-linux-netns

Signed-off-by: Marek Pikuła <m.pikula@partner.samsung.com>

MarekPikula commented on scripts/tests/chiptest/linux.py:
```diff
@@ -70,146 +73,257 @@ def ensure_private_state():
+            dep.setup()
+        self._run_up(self._setup_cmds)
+
+    def activate(self, *args: Any, **kwargs: Any) -> None:
```
That's the point of optional *args

MarekPikula committed:
Restore accidentally removed SubprocessKind.RPC

(and rename it to MGMT to be in line with the new naming)

Signed-off-by: Marek Pikuła <m.pikula@partner.samsung.com>

MarekPikula committed:
Fix small docstring lint

Signed-off-by: Marek Pikuła <m.pikula@partner.samsung.com>

MarekPikula committed:
Fix comment lint

Signed-off-by: Marek Pikuła <m.pikula@partner.samsung.com>

MarekPikula commented on scripts/tests/chiptest/test_definition.py:
```diff
@@ -493,7 +493,7 @@ def _RunImpl(self, target: TestTarget, runner: Runner, apps_register: AppsRegist
             if not dry_run:
                 for key, subproc in subproc_info_repo.items():
                     # Do not add tools to the register
```
Fixed

MarekPikula committed:
Merge remote-tracking branch 'upstream/master' into refactor-linux-netns

MarekPikula committed:
Merge remote-tracking branch 'upstream/master' into refactor-linux-netns

MarekPikula committed:
Fix IPv6 for app and ctrl links

Signed-off-by: Marek Pikuła <m.pikula@partner.samsung.com>

MarekPikula committed:
Fix ns command wrapper for thread router

Signed-off-by: Marek Pikuła <m.pikula@partner.samsung.com>

MarekPikula committed:
Fix wait_for_dad for namespaces

Signed-off-by: Marek Pikuła <m.pikula@partner.samsung.com>

MarekPikula committed:
Fix wait for dad

Signed-off-by: Marek Pikuła <m.pikula@partner.samsung.com>

MarekPikula committed:
Fallback to namespaceless execution in an unlikely corner case

Signed-off-by: Marek Pikuła <m.pikula@partner.samsung.com>

enkiusz commented on scripts/tests/chiptest/linux.py:
```diff
@@ -70,158 +73,273 @@ def ensure_private_state():
+            self._run_cmd(self.down_cmd)
+
+    def _run_cmd(self, command: str) -> None:
+        if isinstance(self.ns_wrapper, str):
```
We should not check class directly, this way an object with __str__ defined can be used here.

```suggestion
        if self.ns_wrapper:
```

enkiusz committed:
Make run_test_suite.py work again, some places were missing --tool-path -> --ctrl-path conversion

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
General improvements on the Linux namespace handling to make it simpler to understand and extend:

- allow for multiple ipv4 / ipv6 addresses to be specified simplifying IPv6 link-local/ULA handling
- dropping the need for _cmd_history by using dependency information to teardown the isolated namespace(s) object graph
- separating out register_dependencies into multiple methods to avoid type matches

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Merge remote-tracking branch 'upstream/master' into MarekPikuła/refactor-linux-netns

enkiusz committed:
Fix ruff errors

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Set accept_ra to work when forwarding is enabled.

MarekPikula commented on scripts/tests/chiptest/linux.py:
```diff
@@ -32,6 +42,271 @@
+class WrappableCmd:
+    """Command used to set up some network resource with optional cleanup command."""
+
+    cmd: str
```
My main goal of having `up` and `down` in (previously) `NetworkCmd` was to be able to track assignment between these two (i.e., this down command relates to this up command), and to have better granularity of them, so that in case of error/cancellation event in the middle of the flow, we don't execute all down commands, but only these, which are required.

MarekPikula commented on scripts/tests/chiptest/linux.py:
```diff
@@ -32,6 +42,271 @@
+            cmd.run(self.ns)
+
+        for dep in self.deps:
+            dep.teardown()
```
If the same dependency is used by multiple network resources the teardown commands will be run multiple times resulting in RuntimeError. Previous approach ensured that down commands are executed only if needed, only once, and only if the resource is no longer used.

MarekPikula commented on scripts/tests/chiptest/linux.py:
```diff
@@ -32,6 +42,271 @@
+    cmd: str
+    ns_wrapper: bool = False
+
+    def run(self, ns) -> None:  # Cannot specify NetworkNamespace here as it's not defined yet
```
Yes you can since we use `from __future__ import annotations`.

MarekPikula commented on scripts/tests/chiptest/linux.py:
```diff
@@ -32,6 +42,271 @@
+    teardown_cmds: list[WrappableCmd] = dataclasses.field(default_factory=list)
+    up_cmds: list[WrappableCmd] = dataclasses.field(default_factory=list)
+    down_cmds: list[WrappableCmd] = dataclasses.field(default_factory=list)
+    deps: list[Self] = dataclasses.field(default_factory=list)
```
I don't think you can use `Self` here. Mypy agrees with me.

MarekPikula reviewed (commented):
Do I understand it correctly that @enkiusz takes over this PR?

My main complaint is the revert of the main feature introduced previously in this PR, namely tight association of up/down commands, and granularization of up/down command flows, allowing for proper cancellation in the middle of the process without errors (or with graceful warning massages in case of down flow failure).

If this PR were to be merged in this state, I would need to submit a follow-up PR reworking the flow (again) to allow for this behavior, as it is a requirement for concurrent tests.

enkiusz committed:
Move network namespace code to matter_testing_infrastructure

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Remove unused import

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Convert to new IsolatedNetworkNamespace API

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Rework state flag checks

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Properly handle check=False in WrappableCmd.run

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Restore SubprocessKind.TOOL and related code (like commandline flags)

Signed-off-by: Maciej Grela <enki@fsck.pl>

enkiusz committed:
Implement post-review suggestions, in particular:

- make networking object ownership explicit, do explicit setup/teardown and drop network resource dependency mechanism
- moved "run in namespace" logic from WrappedCmd to NetworkResource as the namespace is attached there

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Use IsolatedNetworkNamespace abstractions to build otbr-agent commands

Signed-off-by: Maciej Grela <m.grela@samsung.com>

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

arkq commented on src/python_testing/matter_testing_infrastructure/matter/testing/linux/thread.py:
```diff
@@ -22,6 +22,8 @@
 import threading
 from typing import Optional, Pattern
 
+from chiptest.linux import SubprocessKind
```
Import from `chiptest` has to be factored out.

enkiusz committed:
Implement review suggestions, fix wrong import.

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Move SubprocessKind and SubprocessInfo to matter_testing_infrastructure

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Restyled by ruff

enkiusz committed:
Import all symbols used by chiptest.runner's downstream

Signed-off-by: Maciej Grela <m.grela@samsung.com>

restyled-commits committed:
Restyled by isort

enkiusz committed:
Silence ruff check because we need to import these for downstream users

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Don't use python_path to allow scripts to be run outside of venv (looking at you matter_yaml_linter.py)

Signed-off-by: Maciej Grela <m.grela@samsung.com>

restyled-commits committed:
Restyled by isort

enkiusz committed:
Use the `matter.testing.tasks` module instead of `matter.testing.subprocess` not not alias the system `subprocess` module

Signed-off-by: Maciej Grela <m.grela@samsung.com>

restyled-commits committed:
Restyled by isort

enkiusz committed:
More import conversions

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Merge remote-tracking branch 'upstream/master' into refactor-linux-netns

MarekPikula commented on src/python_testing/matter_testing_infrastructure/matter/testing/linux/namespace.py:
```diff
@@ -55,155 +60,251 @@ def ensure_private_state():
+        ],
+            teardown_cmds=[
+            NetworkCmd(f"ip netns del {name}")
+        ])
```
It could be condensed as in NetworkBridge to occupy less lines.

MarekPikula commented on src/python_testing/matter_testing_infrastructure/matter/testing/linux/thread.py:
```diff
@@ -36,17 +38,13 @@ def __init__(self, dataset: str, ns: IsolatedNetworkNamespace):
-        args = [
-            'ip', 'netns', 'exec', self._netns_app, 'otbr-agent', '-d7', '-v', f'-B{self._link_name_app}', radio_url
-        ]
+        cmd = self._netns_app.wrap_cmd(f"otbr-agent -d7 -v '-B{self._link_name_app}' '{radio_url}'")
```
I find this kind of notation very unclean in bash-like manner (i.e., not readable). I think that it would be cleaner if wrap_cmd would accept both `str` (transformed with `shlex.split()`) and `list[str]` and output `list[str]`.

MarekPikula commented on src/python_testing/matter_testing_infrastructure/matter/testing/linux/namespace.py:
```diff
@@ -55,155 +60,251 @@ def ensure_private_state():
+class NetworkLink(NetworkResource):
+    def __init__(self, link_name: str, ipv4_addrs: list[str], ipv6_addrs: list[str], ns: NetworkNamespace | None = None) -> None:
+
+        self.link_name = link_name
```
All network resources have some sort of name, so I think that `name` should be promoted as the `NetworkResource` field.

MarekPikula commented on src/python_testing/matter_testing_infrastructure/matter/testing/linux/namespace.py:
```diff
@@ -55,155 +60,251 @@ def ensure_private_state():
+                "sysctl -w net.ipv6.conf.default.forwarding=1",
+                f"sysctl -w net.ipv6.conf.{self.link_name}.accept_ra=2",
+                f"sysctl -w net.ipv6.conf.{self.link_name}.accept_ra_rt_info_max_plen=64"
+            ]])
```
Why all `ipv6_addrs` conditional additions are spread all over the place? I think that it would be cleaner to first handle ipv4, and then in a single block handle ipv6.

enkiusz committed:
Implement post-review changes:

- install basic python deps in  cert_test_checks.yaml
- use python_path everywhere
- `netns_cmd_wrapper` and `wrap_cmd` now work with list[str] first and foremost with str as fallback
- add documentation for NetworkResource
- consolidate up_cmds for ipv6

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Clean up imports

Signed-off-by: Maciej Grela <m.grela@samsung.com>

restyled-commits committed:
Restyled by isort

enkiusz committed:
Harmonize NetworkLink to use .name like other resource types

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Spelling mistake

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Include system pkgs for matter_yaml_linter.py venv

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Merge remote-tracking branch 'upstream/master' into refactor-linux-netns

enkiusz committed:
Format exception message properly

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Simplify state tracking

Signed-off-by: Maciej Grela <m.grela@samsung.com>

MarekPikula reviewed (commented):
The last nits and we're good.

enkiusz committed:
Promote .name to a property of NetworkResource, properly attach docstrings to fields

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Don't bring up lo if we aren't in a netns, always import matter.testing.tasks using the same package path

Signed-off-by: Maciej Grela <m.grela@samsung.com>

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

enkiusz committed:
Implement review changes:

- always use `with PythonPath` when importing matter_testing_infrastructure
- change up_flag after all up/down commands have been successfull

Signed-off-by: Maciej Grela <m.grela@samsung.com>
